thoth-export-server/src/specification_query.rs

```diff
@@ -0,0 +1,122 @@
+use std::convert::{TryFrom, TryInto};
+use std::sync::Arc;
+use thoth_client::{QueryParameters, ThothClient, Work};
+use thoth_errors::{ThothError, ThothResult};
+use uuid::Uuid;
+
+use crate::record::MetadataSpecification;
+
+enum SpecificationRequest {
+    ByWork,
+    ByPublisher,
+}
+
+pub(crate) struct SpecificationQuery {
+    thoth_client: Arc<ThothClient>,
+    specification: MetadataSpecification,
+}
+
+struct QueryConfiguration {
+    request: SpecificationRequest,
+    specification: MetadataSpecification,
+}
+
+impl SpecificationQuery {
+    pub(crate) fn new(
+        thoth_client: Arc<ThothClient>,
+        specification: MetadataSpecification,
+    ) -> Self {
+        Self {
+            thoth_client,
+            specification,
+        }
+    }
+
+    pub(crate) async fn by_work(self, work_id: Uuid) -> ThothResult<Work> {
+        let parameters: QueryParameters =
+            QueryConfiguration::by_work(self.specification).try_into()?;
+        self.thoth_client.get_work(work_id, parameters).await
+    }
+
+    pub(crate) async fn by_publisher(self, publisher_id: Uuid) -> ThothResult<Vec<Work>> {
+        let parameters: QueryParameters =
+            QueryConfiguration::by_publisher(self.specification).try_into()?;
+        self.thoth_client
+            .get_works(Some(vec![publisher_id]), parameters)
+            .await
+    }
+}
+
+impl QueryConfiguration {
+    fn by_work(specification: MetadataSpecification) -> Self {
+        Self {
+            request: SpecificationRequest::ByWork,
+            specification,
+        }
+    }
+
+    fn by_publisher(specification: MetadataSpecification) -> Self {
+        Self {
+            request: SpecificationRequest::ByPublisher,
+            specification,
+        }
+    }
+}
+
+impl TryFrom<QueryConfiguration> for QueryParameters {
+    type Error = ThothError;
+
+    fn try_from(q: QueryConfiguration) -> ThothResult<Self> {
+        match q.specification {
```

Comment:
Yes, that's a fair point. This change basically allows us to keep having a single struct that we can use (`thoth_client::Work`). The safer option you propose would entail having different queries, with associated structs, so that each specification depended on its particular one. This would indeed be preferred, but would require a major rewrite of both `thoth-client` and `thoth-export-server`...

Reply:
I wasn't thinking of going that far - more like adding some list of fetched/unfetched elements in to each of the modules which would break if not updated by hand, as a reminder to update the `TryFrom`. At the very least, obvious commenting on the `use thoth_client::` statements. It's going to be far too easy to unwittingly break behaviour otherwise, with the `TryFrom` being subtle and separated from the main logic.

Reply:
It's not _that_ different from before, the specification modules have always being agnostic to the query. The best option is to make use of the type system, creating separate queries with associated structs for each specification that needs to be different, when we have the time. For the time being, this is currently breaking the distribution of publisher files, so I'm going to merge it